File: packages/typescript/src/digest.ts
/**
 * SonicOS 7.x Digest auth-int helpers (mirrors Python _auth.py).
 */

import { createHash, randomBytes } from "node:crypto";

export type DigestChallenge = Record<string, string>;

const DIGEST_PARAM_RE = /(\w+)=(?:"([^"]*?)"|([^,\s]+))/g;

export function parseDigestChallenge(wwwAuth: string): DigestChallenge {
  const body = wwwAuth.trim().replace(/^[Dd]igest\s+/, "");
  const params: DigestChallenge = {};
  for (const match of body.matchAll(DIGEST_PARAM_RE)) {
    const key = match[1];
    const value = match[2] ?? match[3] ?? "";
    if (key) params[key] = value;
  }
  return params;
}

function challengePriority(challenge: DigestChallenge): number {
  const alg = (challenge.algorithm ?? "MD5").toUpperCase();
  if (alg === "SHA-256") return 0;
  if (alg === "SHA-256-SESS") return 1;
  return 2;
}

export function pickAuthIntChallenge(
  wwwAuthenticate: string[]
): DigestChallenge | null {
  const candidates: DigestChallenge[] = [];
  for (const value of wwwAuthenticate) {
    if (!value.toLowerCase().startsWith("digest")) continue;
    const parsed = parseDigestChallenge(value);
    if ((parsed.qop ?? "").includes("auth-int")) {
      candidates.push(parsed);
    }
  }
  if (candidates.length === 0) return null;
  return candidates.sort((a, b) => challengePriority(a) - challengePriority(b))[0] ?? null;
}

function hashString(algorithm: string, value: string): string {
  const upper = algorithm.toUpperCase();
  if (upper.includes("SHA-256")) {
    return createHash("sha256").update(value, "utf8").digest("hex");
  }
  return createHash("md5").update(value, "utf8").digest("hex");
}

function hashBytes(algorithm: string, value: Buffer): string {
  const upper = algorithm.toUpperCase();
  if (upper.includes("SHA-256")) {
    return createHash("sha256").update(value).digest("hex");
  }
  return createHash("md5").update(value).digest("hex");
}

export function buildDigestAuthHeader(
  method: string,
  url: string,
  body: Buffer,
  username: string,
  password: string,
  challenge: DigestChallenge
): string {
  const algorithm = (challenge.algorithm ?? "MD5").toUpperCase();
  const realm = challenge.realm ?? "";
  const nonce = challenge.nonce ?? "";
  const opaque = challenge.opaque ?? "";

  const parsed = new URL(url);
  const uri = parsed.pathname + parsed.search;

  const cnonce = randomBytes(8).toString("hex");
  const nc = "00000001";

  let ha1 = hashString(algorithm, `${username}:${realm}:${password}`);
  if (algorithm.includes("SESS")) {
    ha1 = hashString(algorithm, `${ha1}:${nonce}:${cnonce}`);
  }

  const ha2 = hashString(algorithm, `${method}:${uri}:${hashBytes(algorithm, body)}`);
  const response = hashString(algorithm, `${ha1}:${nonce}:${nc}:${cnonce}:auth-int:${ha2}`);

  let header =
    `Digest username="${username}", realm="${realm}", ` +
    `nonce="${nonce}", uri="${uri}", algorithm=${algorithm}, ` +
    `qop=auth-int, nc=${nc}, cnonce="${cnonce}", response="${response}"`;
  if (opaque) header += `, opaque="${opaque}"`;
  return header;
}

export function extractBearerToken(body: unknown): string | null {
  if (!body || typeof body !== "object") return null;
  const info = (body as { status?: { info?: Array<{ bearer_token?: string }> } }).status?.info;
  if (!Array.isArray(info)) return null;
  for (const item of info) {
    if (item?.bearer_token) return String(item.bearer_token);
  }
  return null;
}

File: packages/go/digest.go
package sonicwall

import (
	"crypto/md5"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/url"
	"regexp"
	"strings"
)

var digestParamRE = regexp.MustCompile(`(\w+)=(?:"([^"]*?)"|([^,\s]+))`)

type digestChallenge map[string]string

func parseDigestChallenge(wwwAuth string) digestChallenge {
	body := strings.TrimSpace(wwwAuth)
	body = regexp.MustCompile(`(?i)^digest\s+`).ReplaceAllString(body, "")
	params := digestChallenge{}
	for _, m := range digestParamRE.FindAllStringSubmatch(body, -1) {
		if len(m) < 3 {
			continue
		}
		val := m[2]
		if val == "" {
			val = m[3]
		}
		params[m[1]] = val
	}
	return params
}

func pickAuthIntChallenge(headers []string) digestChallenge {
	var candidates []digestChallenge
	for _, value := range headers {
		if !strings.HasPrefix(strings.ToLower(value), "digest") {
			continue
		}
		c := parseDigestChallenge(value)
		if strings.Contains(c["qop"], "auth-int") {
			candidates = append(candidates, c)
		}
	}
	if len(candidates) == 0 {
		return nil
	}
	best := candidates[0]
	bestPri := digestPriority(best)
	for _, c := range candidates[1:] {
		if p := digestPriority(c); p < bestPri {
			best = c
			bestPri = p
		}
	}
	return best
}

func digestPriority(c digestChallenge) int {
	alg := strings.ToUpper(c["algorithm"])
	switch {
	case alg == "SHA-256":
		return 0
	case alg == "SHA-256-SESS":
		return 1
	default:
		return 2
	}
}

func buildDigestAuthHeader(method, rawURL string, body []byte, username, password string, challenge digestChallenge) string {
	algorithm := strings.ToUpper(challenge["algorithm"])
	if algorithm == "" {
		algorithm = "MD5"
	}
	realm := challenge["realm"]
	nonce := challenge["nonce"]
	opaque := challenge["opaque"]

	parsed, err := url.Parse(rawURL)
	if err != nil {
		parsed = &url.URL{Path: rawURL}
	}
	uri := parsed.Path
	if parsed.RawQuery != "" {
		uri += "?" + parsed.RawQuery
	}

	cnonceBytes := make([]byte, 8)
	_, _ = rand.Read(cnonceBytes)
	cnonce := hex.EncodeToString(cnonceBytes)
	nc := "00000001"

	hashStr := func(s string) string {
		if strings.Contains(algorithm, "SHA-256") {
			sum := sha256.Sum256([]byte(s))
			return hex.EncodeToString(sum[:])
		}
		sum := md5.Sum([]byte(s))
		return hex.EncodeToString(sum[:])
	}
	hashBytes := func(b []byte) string {
		if strings.Contains(algorithm, "SHA-256") {
			sum := sha256.Sum256(b)
			return hex.EncodeToString(sum[:])
		}
		sum := md5.Sum(b)
		return hex.EncodeToString(sum[:])
	}

	ha1 := hashStr(fmt.Sprintf("%s:%s:%s", username, realm, password))
	if strings.Contains(algorithm, "SESS") {
		ha1 = hashStr(fmt.Sprintf("%s:%s:%s", ha1, nonce, cnonce))
	}
	ha2 := hashStr(fmt.Sprintf("%s:%s:%s", method, uri, hashBytes(body)))
	response := hashStr(fmt.Sprintf("%s:%s:%s:%s:auth-int:%s", ha1, nonce, nc, cnonce, ha2))

	header := fmt.Sprintf(
		`Digest username="%s", realm="%s", nonce="%s", uri="%s", algorithm=%s, qop=auth-int, nc=%s, cnonce="%s", response="%s"`,
		username, realm, nonce, uri, algorithm, nc, cnonce, response,
	)
	if opaque != "" {
		header += fmt.Sprintf(`, opaque="%s"`, opaque)
	}
	return header
}

func extractBearerToken(body *SonicOSErrorResponse) string {
	if body == nil {
		return ""
	}
	for _, info := range body.Status.Info {
		if info.BearerToken != "" {
			return info.BearerToken
		}
	}
	return ""
}
